src/Set.ts

```diff
@@ -182,21 +204,24 @@ export const intersection = <A>(S: Setoid<A>): ((x: Set<A>, y: Set<A>) => Set<A>
  * @function
  * @since 1.2.0
  */
-export const partitionMap = <A, L, R>(x: Set<A>, f: (a: A) => Either<L, R>): { left: Set<L>; right: Set<R> } => {
+export const partitionMap = <A, L, R>(x: Set<A>, f: (a: A) => Either<L, R>): Separated<Set<L>, Set<R>> => {
```

Comment:
this is not related to this PR but IMO this signature is wrong, should be 

```ts
export const partitionMap = <L, R>(SL: Setoid<L>, SR: Setoid<R>) => <A>(x: Set<A>, f: (a: A) => Either<L, R>): Separated<Set<L>, Set<R>>
```

Reply:
But why? I was also wondering why Set needs a Setoid to check the existence in current implementation. JavaScript Set implementation already filters out duplicates so you can't create `new Set([1, 1])`, result will always be `new Set([1])`. The same applies for `Set.prototype.add`

Reply:
Furthermore `Filterable` instance requires `Set` to be a valid `Functor` without additional `Setoid`'s. And all tests pass for `map` without `Setoid`

Reply:
Because equality is a serious matter, actually as a mathematician I think that equality is the only thing that really matters :)

`Set` uses implicitly the following setoid

```ts
const setoidEvil: Setoid<any> = {
  equals: (x, y) => x === y
}
```

which is useless, try this `new Set([{ a: 1 }, { a: 1 }])`. 

Actually is worse than useless, is evil, because if you use that setoid, you are pretty much out of funcitonal programming (unless you add some machinery everytime you define a struct/function).

Let's define a `Person` struct

```ts
interface Person {
  name: string
}

const p1: Person = { name: 'foo' }
const p2: Person = { name: 'foo' }
```

"obviously" `p1` and `p2` are equal, however

```ts
console.log(setoidEvil.equals(p1, p2)) // => false
```

Even worse, consider the following function `person`

```ts
// constructor for Person
const person = (name: string): Person => ({ name })

const x1 = 'foo'
const x2 = 'foo'
const y1 = person(x1)
const y2 = person(x2)

console.log(setoidEvil.equals(x1, x2)) // => true
// but...
console.log(setoidEvil.equals(y1, y2)) // => false
```

as you can see `person` is not a pure function, which is (in practice) absurd.

So basically everytime you define an interface or a function or whatever, you are also defining an implicit setoid (or more setoids) instance.
Normally keeping those setoids implicit isn't much of a problem, but in the case of `Set` they are important and we should make them explicit, otherwise you can get duplicates (or even worse results).

Reply:
Totally makes sense!
What about implementing it similarly to `Either`/`Validation` and accept a setoid in `getCompactable`?

Reply:
@gcanti I've updated `Set`, take a look

Reply:
@gcanti What do you think about such approach in general? When you can define all type arguments for a typeclass? I mean that `CompactableA`/`Compactable1A` which is similar to `Compactable2C`. What do you think about it? Is it ok? Is naming appropriate?

Reply:
I'm not sure, it doesn't feel right (and there's no equivalent in purescript-filterable). 
Likely there's no `Compactable` instance for `Set` just as there's no `Functor` instance.
For now I propose to either 

- remove `Set` from this PR and rediscuss later
- or remove `getCompactable` and define `compact` and `separate` as

```ts
export const compact = <A>(S: Setoid<A>) => (fa: Set<Option<A>>): Set<A> => ...
export const separate = <A, B>(SA: Setoid<A>, SB: Setoid<B>) => (fa: Set<Either<A, B>>): Separated<Set<A>, Set<B>> => ...
```

Reply:
Ok, let's revert Set now and come back to it later.

Reply:
@gcanti Ready